src/radix_connect/mobile/deep_link_parsing/request.rs

```diff
@@ -0,0 +1,162 @@
+use crypto::signatures::ed25519::Signature;
+
+use super::super::session::session_id::SessionID;
+use crate::prelude::*;
+use hex::ToHex;
+
+#[derive(Debug, PartialEq)]
+pub struct RadixConnectMobileRequest {
+    pub session_id: SessionID,
+    pub origin: DappOrigin,
+    pub public_key: KeyAgreementPublicKey,
+    pub identity_public_key: Ed25519PublicKey,
+    pub dapp_definition_address: DappDefinitionAddress,
+    pub signature: Ed25519Signature,
+    pub interaction: DappToWalletInteractionUnvalidated,
+}
+
+impl RadixConnectMobileRequest {
+    pub fn new(
+        session_id: SessionID,
+        origin: DappOrigin,
+        public_key: KeyAgreementPublicKey,
+        identity_public_key: Ed25519PublicKey,
+        dapp_definition_address: DappDefinitionAddress,
+        signature: Ed25519Signature,
+        interaction: DappToWalletInteractionUnvalidated,
+    ) -> Self {
+        Self {
+            session_id,
+            origin,
+            public_key,
+            identity_public_key,
+            dapp_definition_address,
+            signature,
+            interaction,
+        }
+    }
+}
+
+impl HasSampleValues for RadixConnectMobileRequest {
+    fn sample() -> Self {
+        RadixConnectMobileRequest::new(
+            SessionID::sample(),
+            DappOrigin::sample(),
+            KeyAgreementPublicKey::sample(),
+            Ed25519PublicKey::sample(),
+            DappDefinitionAddress::sample(),
+            Ed25519Signature::sample(),
+            DappToWalletInteractionUnvalidated::sample(),
+        )
+    }
+
+    fn sample_other() -> Self {
+        RadixConnectMobileRequest::new(
+            SessionID::sample(),
+            DappOrigin::sample_other(),
+            KeyAgreementPublicKey::sample_other(),
+            Ed25519PublicKey::sample_other(),
+            DappDefinitionAddress::sample(),
+            Ed25519Signature::sample(),
+            DappToWalletInteractionUnvalidated::sample(),
+        )
+    }
+}
+
+impl RadixConnectMobileRequest {
+    pub fn verify_request_signature(&self) -> Result<()> {
+        let message = self.message_for_signature();
+        self.verify_message_signature(&message)
+    }
+
+    pub fn verify_message_signature(&self, message: &Hash) -> Result<()> {
+        let is_valid_signature = self
+            .identity_public_key
+            .is_valid_signature_for_hash(&self.signature, message);
+
+        if is_valid_signature {
+            Ok(())
+        } else {
+            Err(CommonError::RadixConnectMobileInvalidDappSignature)
+        }
+    }
+
+    fn message_for_signature(&self) -> Hash {
+        dapp_message_for_signature(
+            &self.interaction.interaction_id,
+            &self.dapp_definition_address,
+            &self.origin,
+        )
+    }
+}
+
+pub fn dapp_message_for_signature(
```

Comment:
I would split `dapp_message_for_signature` into `dapp_message_for_signature` and `hashed_dapp_message_for_signature` I think.

And have to separate tests for them.